### ISSUE TITLE
TFLite GPU execution failed

### ISSUE THREAD

**Issue by joyalbin** [tensorflow/tensorflow#35632]
I trying to run one model on TFite mobile GPU using opencl and getting below error:

> INFO: Initialized TensorFlow Lite runtime.                                                                                           
> INFO: Created TensorFlow Lite delegate for GPU.                                                                                      
> ERROR: Next operations are not supported by GPU delegate:                                                                            
> SPLIT_V: Operation is not supported.                                                                                                 
> First 294 operations will run on the GPU, and the remaining 40 on the CPU.                                                           
> INFO: Initialized OpenCL-based API.                                                                                                  
> Applied GPU delegate.                                                                                                                
> Initialized session in 107237ms.                                                                                                     
> Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.                         
> ERROR: TfLiteGpuDelegate Invoke: Failed to read data from GPU (clEnqueueReadBuffer) - Execution status error for events in wait list 
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.                                                                       
>                                                                                                                                      
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.                                                                                                            
>                                                                                                                                      
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.
> 
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.
> 
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.
> 
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.
> 
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.
> 
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.
> 
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.
> 
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.
> ERROR: TfLiteGpuDelegate Invoke: Failed to upload data to GPU (clEnqueueWriteBuffer) - Execution status error for events in wait list
> ERROR: Node number 334 (TfLiteGpuDelegateV2) failed to invoke.
> 
> count=138 first=73906 curr=1840 min=1414 max=73906 avg=3142.52 std=6057
> 

Please help me to resolve this issue

**Comment by aselle** [comment]
Can you specify more details about what device you are trying this on? are you using a standard model?

**Comment by impjdi** [comment]
Is it possible to share the model?

**Comment by joyalbin** [comment]
@aselle @impjdi I am using a standard model. My model contain Conv2D=>InstanceNorm=>relu. The problem is coming from instancenorm implementation.

InstanceNorm is implemented with combination of operators since Tensorflow doesn't have a single operator for this.
I have used Matmul to implement the reduce_mean to calculate Mean and Variance part of InstanceNorm. 

I am running on mobile phone using tflite GPU. Tflite CPU its working fine

**Comment by joyalbin** [comment]
I am also sharing the part of the model with I am facing the problem. Please help me to find the reason for the issue and any alternative solution?

[conv2d-instancenorm-relu.zip](https://github.com/tensorflow/tensorflow/files/4034657/conv2d-instancenorm-relu.zip)

**Comment by impjdi** [comment]
Thank you for the network.  I forwarded the issue to our OpenCL specialist.

**Comment by impjdi** [comment]
He has confirmed that https://github.com/tensorflow/tensorflow/files/4034657/conv2d-instancenorm-relu.zip works fine on his device.  What kind of device are you using?

**Comment by joyalbin** [comment]
@impjdi Thanks for your reply.
I am trying to run this on mobile device Honor20. Could you also please let me know in which environment it works fine?

**Comment by nerdaliciousCH** [comment]
Same on LG Nexus 5X

Running posenet_mobilenet_v1_100_257x257_multi_kpt_stripped.tflite downloaded from the tensorflow lite webpage just days ago.
With CPU it works pefectly fine

```
 java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: OpenCL library not loaded - dlopen failed: library "libOpenCL-pixel.so" not found
    Falling back to OpenGL
    TfLiteGpuDelegate Invoke: ToTensorConverter: input data size does not match expected size.
    Node number 31 (TfLiteGpuDelegateV2) failed to invoke.
    
    TfLiteGpuDelegate Invoke: ToTensorConverter: input data size does not match expected size.
    Node number 31 (TfLiteGpuDelegateV2) failed to invoke.
    
        at org.tensorflow.lite.NativeInterpreterWrapper.run(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:171)
        at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:275)
```

**Comment by impjdi** [comment]
@joyalbin Apologies, forgot to follow up on this.  He usually tests it on a variety of phones.  I don't think we have Honor20 in particular but definitely have Kirin devices.  Is this now resolved or shall I ping him again?

**Comment by impjdi** [comment]
@nerdaliciousCH the way I read the code is that you are supplying an external SSBO for inference, but it has wrong dimensions, i.e. not 257x257x3.

**Comment by nerdaliciousCH** [comment]
> @nerdaliciousCH the way I read the message is that you are supplying an external SSBO for inference, but it has wrong dimensions, i.e. not 257x257x3.

So I use Android with Java and used the GpuDelegate provided by the tensorflow2 packages. When I don't add the delegate to the Intepreter and run on CPU alone (i.e. nlt providing any executor delegates), it works well. I have no code that is adapted to the device type besides when I create the tensorflow interpreter.

Do I have to create the input buffers differently when running on GPU? What struck me as odd is the OpenCL related part of the Exception. Could the size error be a consequence of that somewhere downstream the call stack?

**Comment by impjdi** [comment]
@nerdaliciousCH 
> Do I have to create the input buffers differently when running on GPU?
No, you don't have to create the input buffers differently.  In fact, it should be just the same.
> What struck me as odd is the OpenCL related part of the Exception. 
Don't worry about OpenCL.  It's just a "FYI" message.
> Could the size error be a consequence of that somewhere downstream the call stack?
Maybe, but it's hard to tell without seeing the code.  Are you calling interpreter->ResizeInputTensors at one point?

**Comment by bilalsoomro** [comment]
Hi. I also have a basic CNN model that is crashing with a similar error only when using GPU delegate (Runs fine on CPU and with NNAPI Delegate). 

Strangely it only crashes on the Samsung S9 (Android 10) with a Exynos 9810. It doesn't crash on any other device that I have tested (Kirin and Snapdragon). I am using the Java TFLite API.

Here's a link to the model: [link](https://drive.google.com/file/d/1HnclW3Qh_LOkIjMrd5gd9VfShITMbhwL/view?usp=sharing)

Here's the error message:
```
2020-03-04 09:34:06.950 16795-16795/com.test.app E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.test.app, PID: 16795
    java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: TfLiteGpuDelegate Invoke: Failed to read data from GPU (clEnqueueReadBuffer) - Execution status error for events in wait list
    Node number 8 (TfLiteGpuDelegateV2) failed to invoke.
    
        at org.tensorflow.lite.NativeInterpreterWrapper.run(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:154)
        at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:314)
        at org.tensorflow.lite.Interpreter.run(Interpreter.java:275)
```

**Comment by nerdaliciousCH** [comment]
@impjdi 
No resizing tensors in my code :(

**Comment by impjdi** [comment]
@bilalsoomro Please sync to the latest git commit and try it out.  A fix should have been submitted.

**Comment by bilalsoomro** [comment]
Hi @impjdi, the latest build has fixed the issue.

**Comment by OleksandrGrument** [comment]
Hello, have the same issue. All works fine on different devices except of Samsung. Model S730FM.
Exynos 7 Octa (Exynos 7870) 
GPU - Mali-T830

Tried on different gpus: Adreno 615, Adreno 508, Adreno 509, Mali - G72, Adreno 540. Works well.

Tensorflow 2.3.0 

` something went wrong: Internal error: Failed to run on the given Interpreter: Following operations are not supported by GPU delegate:
    SUM: Operation is not supported.
    226 operations will run on the GPU, and the remaining 103 operations will run on the CPU.
    TfLiteGpuDelegate Invoke: Failed to clEnqueueNDRangeKernel - Out of resources
    Node number 329 (TfLiteGpuDelegateV2) failed to invoke.`

When i run on cpu without gpu delegate all code works well.

If anyone have idea how to handle it? Maybe there is some trick how to understand that gpu not support operations without running inference? If there a possibility check gpu compatability ? 

According to this articles, Mali-T830 does'nt have support of opencl 2.0. Looks like possible reason of this behaviour.
https://en.wikipedia.org/wiki/Mali_(GPU)
https://en.wikipedia.org/wiki/Adreno

**Comment by BilalAndroidDev** [comment]
> Same on LG Nexus 5X
> 
> Running posenet_mobilenet_v1_100_257x257_multi_kpt_stripped.tflite downloaded from the tensorflow lite webpage just days ago.
> With CPU it works pefectly fine
> 
> ```
>  java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: OpenCL library not loaded - dlopen failed: library "libOpenCL-pixel.so" not found
>     Falling back to OpenGL
>     TfLiteGpuDelegate Invoke: ToTensorConverter: input data size does not match expected size.
>     Node number 31 (TfLiteGpuDelegateV2) failed to invoke.
>     
>     TfLiteGpuDelegate Invoke: ToTensorConverter: input data size does not match expected size.
>     Node number 31 (TfLiteGpuDelegateV2) failed to invoke.
>     
>         at org.tensorflow.lite.NativeInterpreterWrapper.run(Native Method)
>         at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:171)
>         at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:275)
> ```

java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: OpenCL library not loaded - dlopen failed: library "libOpenCL-pixel.so" not found Falling back to OpenGL TfLiteGpuDelegate Invoke: ToTensorConverter: input data size does not match expected size. Node number 237 (TfLiteGpuDelegateV2) failed to invoke. 

I'm facing the same error on Pixel 2 XL (Android 11).

Actually, it was working fine with Android 10 on the same device but getting the same mentioned issue when I updated it to Android 11.

I'm using the following libraries for tensoreflow:
org.tensorflow:tensorflow-lite:2.0.0
org.tensorflow:tensorflow-lite-gpu:2.0.0

Is anyone got success to get rid of this issue? @nerdaliciousCH

**Comment by tgpsantos** [comment]
Hi guys, I'm getting the same error in a new model that I'm testing in [Huawei P10 Lite LX1A](https://www.gsmarena.com/huawei_p10_lite-8598.php). In Samsung S9 it worked fine.

The P10 Lite has a **Mali-T830** GPU, like @OleksandrGrument mentioned above. I'm also running with `Tensorflow 2.3.0`

**Comment by impjdi** [comment]
This thread is diluted with multiple reports.  Can I get a TFLite model that reproduces this?  We'll see whether we have a Mali-T830 device at hand.  (FYI Due to COVID-19 closed offices, we don't have access to the devices anymore and may not be able to try out that particular (or a similar) device.)

**Comment by tgpsantos** [comment]
@impjdi , [here it is](https://we.tl/t-GOEGyHTmqS). Thank you in advance

**Comment by impjdi** [comment]
@tgpsantos I've verified that your model works fine on a Pixel 4 OpenGL and OpenCL.  I'll ping the team whether anyone can reproduce this on any of their machines they have.

**Comment by impjdi** [comment]
@tgpsantos A teammate ran your model on a Huawei P9 (Mali T880MP4) and it worked fine, matching the CPU output.  Can you sync to the latest master branch and double-check?

**Comment by tgpsantos** [comment]
Updated to from 2.3.0 to 2.4.0 and problem was solved. :) Thank you @impjdi and the team.

PS: I had no idea there was a new tflite release. Are they in-sync with python TF releases? If not, where can I learn about new releases?

**Comment by impjdi** [comment]
TFLite GPU team is not really familiar with the different TF releases really; we work with the master branch's latest code only.

**Comment by walterdumba** [comment]
> > Same on LG Nexus 5X
> > Running posenet_mobilenet_v1_100_257x257_multi_kpt_stripped.tflite downloaded from the tensorflow lite webpage just days ago.
> > With CPU it works pefectly fine
> > ```
> >  java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: OpenCL library not loaded - dlopen failed: library "libOpenCL-pixel.so" not found
> >     Falling back to OpenGL
> >     TfLiteGpuDelegate Invoke: ToTensorConverter: input data size does not match expected size.
> >     Node number 31 (TfLiteGpuDelegateV2) failed to invoke.
> >     
> >     TfLiteGpuDelegate Invoke: ToTensorConverter: input data size does not match expected size.
> >     Node number 31 (TfLiteGpuDelegateV2) failed to invoke.
> >     
> >         at org.tensorflow.lite.NativeInterpreterWrapper.run(Native Method)
> >         at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:171)
> >         at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:275)
> > ```
> 
> java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: OpenCL library not loaded - dlopen failed: library "libOpenCL-pixel.so" not found Falling back to OpenGL TfLiteGpuDelegate Invoke: ToTensorConverter: input data size does not match expected size. Node number 237 (TfLiteGpuDelegateV2) failed to invoke.
> 
> I'm facing the same error on Pixel 2 XL (Android 11).
> 
> Actually, it was working fine with Android 10 on the same device but getting the same mentioned issue when I updated it to Android 11.
> 
> I'm using the following libraries for tensoreflow:
> org.tensorflow:tensorflow-lite:2.0.0
> org.tensorflow:tensorflow-lite-gpu:2.0.0
> 
> Is anyone got success to get rid of this issue? @nerdaliciousCH @impjdi


Hi there,

I was experiencing the same issue with the following configuration:

OnePlus One
Adreno 330
Android 10
org.tensorflow:tensorflow-lite:2.4.0
org.tensorflow:tensorflow-lite-gpu:2.4.0

I downgrade to 2.1.0 and worked

**Comment by hamlatzis** [comment]
I have the same error, but cannot update to newer TF version as I've made modifications and aren't compatible to the newer branch.

Can you share the difference between the two branches (2.3.0 & 2.4.0) that fix this error, maybe I could incorporate them in my version?

thnx

**Comment by gowthamkpr** [comment]
@hamlatzis @walterdumba Can you guys check with the latest versions of tensorflow and see if the issue still persists. Thanks!

**Comment by mohantym** [comment]
Hi @joyalbin !
Could you check with TF lite 2.9/2.10/nightly version. There has lot of bug fixes in Tensorflow and TF Lite since 2.8 version.
Thank you!
